csrc/scheduler/mma_utils.cpp

```diff
@@ -1397,6 +1400,88 @@ MatmulProblemLayoutOpt getMatmulLayout(Fusion* fusion) {
   return {"Failed to decide fusion inputs' data layout."};
 }
 
+RolesMapOpt getTensorsRoles(Fusion* fusion) {
+  ComputeAtMap ca_map(fusion);
```

Comment:
nit: Don't have to be in this PR, but we are building the CA map multiple times. We should build it once, and use it everywhere. For example:
https://github.com/NVIDIA/Fuser/blob/37f45332e6890511628ea3e660213d1b4d0340f1/csrc/scheduler/pointwise.cpp#L89-L93

Reply:
Thanks for pointing this out.

When I prepared this PR I wasn't sure how expensive it is to create CA map but as you wrote, it would be best to have it created as few times as it is possible, ideally one instance perf fusion.

I checked execution paths and we need CA map in two places:
- compile time check, we need it for calculating layout and checking if core roles are detected in fusion inputs/outputs
- `scheduleMatmul`, here we need to access tensors based on their roles and to infer the layout of fusion

The issue is that we currently cannot create a `HeuristicSummaryEntry` instance when we run compile-time checks on fusion (no access to `data_cache` instance) so created instance of CA map cannot be cached. It wouldn't be also too useful as heuristic cache is not available in `schedule` of scheduler.
The idea we could follow here is dependency injection so we can reduce the number of created instances to at most two, used in places mentioned above. The drawback is that API will be less clear, by passing `const&` or `const std::unique_ptr<ComputeAtMap>&` parameters.